File: server/lib/jd-fetcher.js
import * as cheerio from "cheerio";

const HEADERS = {
  "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
};

export async function fetchJdFromUrl(url) {
  if (!url.startsWith("http://") && !url.startsWith("https://")) {
    throw new Error("URL must start with http:// or https://");
  }

  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 15000);

  let resp;
  try {
    resp = await fetch(url, { headers: HEADERS, signal: controller.signal });
  } catch (err) {
    if (err.name === "AbortError") throw new Error("URL timed out after 15 seconds");
    throw new Error("Could not connect to that URL");
  } finally {
    clearTimeout(timeout);
  }

  if (!resp.ok) {
    throw new Error(`URL returned HTTP ${resp.status}`);
  }

  const html = await resp.text();
  const $ = cheerio.load(html);

  $("script, style, nav, footer, header").remove();

  const main =
    $("article").first() ||
    $("main").first() ||
    $('[class*="job-description"], [class*="job_description"], [class*="description"], [class*="content"], [class*="posting"]').first() ||
    $("body").first();

  let text = (main.length ? main.text() : $.root().text()) || "";
  const lines = text
    .split("\n")
    .map((l) => l.trim())
    .filter((l) => l);
  const cleaned = lines.join("\n");
  return cleaned.slice(0, 6000);
}

File: backend/utils/jd_fetcher.py
import requests
from bs4 import BeautifulSoup

HEADERS = {"User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"}


def fetch_jd_from_url(url: str) -> str:
    if not url.startswith(("http://", "https://")):
        raise ValueError("URL must start with http:// or https://")

    try:
        resp = requests.get(url, headers=HEADERS, timeout=15)
        resp.raise_for_status()
    except requests.Timeout:
        raise RuntimeError("URL timed out after 15 seconds")
    except requests.ConnectionError:
        raise RuntimeError("Could not connect to that URL")
    except requests.HTTPError as e:
        raise RuntimeError(f"URL returned HTTP {e.response.status_code}")

    soup = BeautifulSoup(resp.text, "html.parser")

    for tag in soup(["script", "style", "nav", "footer", "header"]):
        tag.decompose()

    main = (
        soup.find("article") or
        soup.find("main") or
        soup.find(attrs={"class": lambda c: c and any(
            k in str(c).lower() for k in ["job-description", "job_description", "description", "content", "posting"]
        )}) or
        soup.find("body")
    )

    text = main.get_text(separator="\n", strip=True) if main else soup.get_text(separator="\n", strip=True)
    lines = [line.strip() for line in text.splitlines() if line.strip()]
    cleaned = "\n".join(lines)
    return cleaned[:6000]
